### PR TITLE
[Docs] Add accelerator backend test reuse guide

### PR DESCRIPTION
**Pull request by orangeH25**
## Summary

This PR adds a guideline document for out-of-tree (OOT) accelerator backend developers on reusing PyTorch's upstream test suite.

The guide covers:

- Test categorization and the role of hardware classification in selecting applicable tests.
- Test reuse strategies for OOT backends, including adapting upstream tests to match backend capabilities without modifying upstream tests.
- Configuration options for handling unsupported functionality, expected backend differences, and gradually expanding test coverage.
- A recommended bring-up workflow for incrementally adopting upstream tests.

### CODE REVIEW

File: docs/source/accelerator/testing.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+| Classification                 | Meaning                                      |
+| ------------------------------ | -------------------------------------------- |
+| `GENERIC`                      | No accelerator needed — pure framework logic |
+| `DEVICE_GENERIC`               | Expected to run on all accelerator backends  |
+| `CPU` / `CUDA` / `MPS` / `XPU` | Specific device type (use sparingly)         |
+
+
+### Generic Tests
+
+Generic tests validate framework functionality that does not depend on accelerator support. OOT backends can run these tests directly in their PyTorch environment:
```
Comment (albanD) [inline docs/source/accelerator/testing.md]: Actually shouldn't these never be ran by OOT backends. Core CI does run this for all supported CPU arch already ?

File: docs/source/accelerator/testing.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+### Device-specific Tests
+
+Device-specific tests target functionality or behavior tied to a particular device type (e.g. `CPU`, `CUDA`, `MPS`, `XPU`):
+
+```bash
+python test/run_test.py --hw-classification CUDA
+```
+
+If an OOT backend provides equivalent functionality, applicable tests may be reused through the normal device test instantiation flow:
```
Comment (albanD) [inline docs/source/accelerator/testing.md]: What does this one mean?

File: docs/source/accelerator/testing.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+- Backend-specific numerical behavior or precision differences.
+
+OOT backends can adapt upstream tests through the configuration options described in [Test Reuse Configuration](#test-reuse-configuration) below, rather than forking and modifying the tests.
+
+### Device-specific Tests
+
+Device-specific tests target functionality or behavior tied to a particular device type (e.g. `CPU`, `CUDA`, `MPS`, `XPU`):
+
+```bash
+python test/run_test.py --hw-classification CUDA
```
Comment (albanD) [inline docs/source/accelerator/testing.md]: I guess this would be a noop for OOT backends?

But we should have a PU1 one? That contains the basic PU1 logic validation test in core that check if PU1 backend was registered properly?

File: docs/source/accelerator/testing.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+### Restricting Test Scope
+
+Test scope restriction allows OOT backends to reuse upstream tests selectively based on their supported functionality. Unsupported cases can be excluded and test behavior can be customized through the following configurations:
+
+| Configuration     | Type              | Purpose                                                      |
+| ----------------- | ----------------- | ------------------------------------------------------------ |
+| `op_allowlist`    | `Collection[str]` | Limit `@ops` tests to a supported operator subset            |
+| `test_exclusions` | `dict`            | Exclude unsupported test classes, methods, or dtype variants |
+
+> These configurations can be set through `set_test_configs()`. Calling `set_test_configs()` without arguments resets all configurations to `None`.
```
Comment (albanD) [inline docs/source/accelerator/testing.md]: Can you provide the full path?

File: docs/source/accelerator/testing.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+#### `op_allowlist`
+
+Limit `@ops`-parametrized tests to supported operators. Values match `OpInfo.full_name` (e.g. `"add.Tensor"`, `"linalg.norm"`):
+
+```python
+PrivateUse1TestBase.set_test_configs(
```
Comment (albanD) [inline docs/source/accelerator/testing.md:67]: Can you clarify if it's ok to call this multiple times? Or every config should be in a single call

File: docs/source/accelerator/testing.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+)
+```
+
+Overrides are applied on top of existing `OpInfo` decorators. When both `op_allowlist` and `op_overrides` are set, `op_allowlist` filters first, then `op_overrides` is applied to the remaining operators.
+
+
+
+### Expanding Test Scope
+
+#### `bypass_device_restrictions`
```
Comment (albanD) [inline docs/source/accelerator/testing.md:115]: Should we state here that users should send PRs to core to update?

File: docs/source/accelerator/testing.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    bypass_device_restrictions = True
+
+instantiate_device_type_tests(TestPrimsOpenReg, globals(), only_for="openreg")
+```
+
+
+
+## Bring-up Workflow
+
+A recommended order for adopting PyTorch tests in a new OOT backend:
```
Comment (albanD) [inline docs/source/accelerator/testing.md]: I think there is some details about which test file to run and how this would apply

File: docs/source/accelerator/testing.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+1. **Identify reusable tests** — use `--hw-classification GENERIC DEVICE_GENERIC` to find tests that may be reused.
+2. **Restrict unsupported ops** — use `op_allowlist` to run only supported operators during early bring-up.
+3. **Handle backend differences** — add `op_overrides` for expected failures, skips, and precision differences.
+4. **Exclude incompatible tests** — use `test_exclusions` only for functionality that cannot reasonably be reused.
+
+> If a reusable test is blocked by legacy device guards (such as `onlyCUDA`), set `bypass_device_restrictions = True` when the backend supports the required functionality.
+
+For a complete runnable example, see [test_testing.py][OpenReg test_testing].
```
Comment (albanD) [inline docs/source/accelerator/testing.md]: Would this link actually work?

File: docs/source/accelerator/testing.md (hunk 1 above)
Comment (orangeH25) [reply docs/source/accelerator/testing.md]: Agreed, thanks! OOT backends indeed don't need to run these tests. I'll update the doc to clarify this.

File: docs/source/accelerator/testing.md (hunk 2 above)
Comment (orangeH25) [reply docs/source/accelerator/testing.md]: Sorry, I think I was looking at this from the perspective of the ongoing migration rather than the final classification state. After the classification is fully rolled out, tests classified as `CUDA` should represent real CUDA-specific dependencies and should not be reused by OOT backends. I'll remove this part. Thanks!

File: docs/source/accelerator/testing.md (hunk 3 above)
Comment (orangeH25) [reply docs/source/accelerator/testing.md]: Agreed that `--hw-classification CUDA`/`MPS`/`XPU` are no-ops for OOT backends — I'll drop them from this doc, they're not relevant to the OOT audience. 
For the PU1 registration check, I think it fits best as step 0 in the Bring-up Workflow section — before running any test suite, the backend developer should first verify their registration is wired up correctly. Something like:

```python
import torch
# Verify PU1 backend registration
torch.accelerator.current_accelerator()   	# should return your device
torch.accelerator.device_count()          	# should be > 0
torch.tensor([1], device="your_backend")   	# basic tensor creation
```

File: docs/source/accelerator/testing.md (hunk 4 above)
Comment (orangeH25) [reply docs/source/accelerator/testing.md]: Sure, I'll update it.

File: docs/source/accelerator/testing.md (hunk 5 above)
Comment (orangeH25) [reply docs/source/accelerator/testing.md:67]: Good point, thanks! I intended to imply this behavior with the note that calling `set_test_configs()` without arguments resets all configurations to `None`, but that it is not clear. I'll clarify that `set_test_configs()` replaces the current configuration and all desired configs should be provided in a single call.

File: docs/source/accelerator/testing.md (hunk 6 above)
Comment (orangeH25) [reply docs/source/accelerator/testing.md:115]: Good point, I'll add a note here as a reminder, thanks!

File: docs/source/accelerator/testing.md (hunk 7 above)
Comment (orangeH25) [reply docs/source/accelerator/testing.md]: Yes, I'll improve this section. Thanks!

File: docs/source/accelerator/testing.md (hunk 8 above)
Comment (orangeH25) [reply docs/source/accelerator/testing.md]: Yes, it should work. It's currently using reference-style links, I'll switch it to inline links.

**Comment by fffrog** [comment]
Great work @orangeH25, thank you very much for contributing this one.

This is just my personal opinion on this document; feel free to discuss it with me at any time.

**Background**

PyTorch has a large upstream test suite. OOT backends can reuse these tests directly to ensure quality. Two mechanisms make this possible: **test classification labels** and **dynamic config/filtering**. No upstream code changes needed.

**How it works**

We need a dedicated section to show how it works and the overall desigin and principle.

1. `HardwareClassification` — labels test classes (`GENERIC` / `ACCELERATOR` / device-specific). `HardwareClassificationTestLoader` filters tests at discovery time via `--hw-classification`.

2. `instantiate_device_type_tests` — expands a template TestClass into device-specific classes like `TestFooCPU` / `TestFooCUDA` / `TestFooPrivateUse1`. During expansion, it reads config attributes (`test_exclusions` / `op_allowlist` / `op_overrides`) from `DeviceTypeTestBase` (OOT backends use `PrivateUse1TestBase`) to do dynamic filtering. Two env vars are supported:
   - `PYTORCH_TESTING_DEVICE_ONLY_FOR`: limit which device types to run
   - `PYTORCH_TESTING_DEVICE_FOR_CUSTOM`: map some CUDA tests to custom device

**How to use**

From the developers perspective, I think the most important things should be `Test classification` and `Dynamic configaration`, so we need to introduce them one by one in details.

**Test classification:**
- `GENERIC`: pure logic, runs on CPU, optional for OOT backends
- `ACCELERATOR`: needs an accelerator but logic is device-agnostic — **highest priority for OOT backends**
- `CPU` / `CUDA` / `XPU`: device-specific only. OOT backends should put their own device-specific tests in their downstream repo, inheriting from `PrivateUse1TestBase`

Give the example of how to use it.

**Dynamic configuration:**

It would be better if we can expand those features with the detailed explanation and examples

| Mechanism | What it does | When to use |
|-----------|-------------|-------------|
| `op_allowlist` | Operator allowlist | Early bring-up |
| `test_exclusions` | Exclude class/method/dtype-level variants | Unsupported features (dtype exclusion only works for `@dtypes`/`@dtypesIf`) |
| `op_overrides` | Per-operator decorator overrides | Specific ops need special handling |
| `bypass_device_restrictions` | Skip `@onlyCUDA` and similar guards | Temporary workaround |

**Bring-up workflow**

Give an example to illustrate how to use all things together.

**Suggestions**

- Connect downstream repo to upstream CI via CRCR (see `ci.md`)
- Per PR CI should run `ACCELERATOR` tests first

**Comment by orangeH25** [comment]
Thanks @fffrog , this looks much clearer. Really appreciate the suggestions! I'll update the document accordingly and expand the sections with more details and examples.

File: docs/source/accelerator/testing.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+
+**For OOT backends, upstream test reuse should primarily focus on `ACCELERATOR` tests.** `GENERIC` tests are already covered by upstream CPU CI. Device-specific tests target built-in backends. OOT backends typically maintain their own device-specific tests downstream.
+
+
+**Running upstream `ACCELERATOR` tests**
+
+Configure the following environment variables before running the upstream test suite:
+
+- `PYTORCH_TESTING_DEVICE_FOR_CUSTOM` adds the specified custom device type(s) for testing.
```
Comment (fffrog) [inline docs/source/accelerator/testing.md:28]: I think this one is not necessary

File: docs/source/accelerator/testing.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
+# Test Suite Reuse
+
+## Background
+
+PyTorch has a large upstream test suite. OOT backends can reuse these tests directly to ensure quality. Two mechanisms make this possible: **test classification** and **dynamic test configuration**. Without requiring changes to upstream tests.
+
+## Core Mechanisms
+
+### Test Classification
```
Comment (KarhouTam) [inline docs/source/accelerator/testing.md:9]: What if OOT backends have their own test classes? Those also need to be tagged. I think we should add a short code snipet about how to tag a test class here.

File: docs/source/accelerator/testing.md [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+| Option | What it does | When to use |
+|---|---|---|
+| `op_allowlist` | Operator test allowlist | Early bring-up |
+| `test_exclusions` | Exclude class/method/dtype-level variants | Unsupported features |
+| `op_overrides` | Per-operator decorator overrides | Specific ops need special handling |
+| `bypass_device_restrictions` | Skip `@onlyCUDA` and similar guards | Temporary workaround |
+
+OOT backends configure all options except `bypass_device_restrictions` through `PrivateUse1TestBase.set_test_configs()` (inherited from [`DeviceTypeTestBase`](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_device_type.py)). 
+
+Each call to `set_test_configs()` fully replaces the existing configuration: omitted keyword arguments are reset to `None`. Therefore, provide all desired configurations in a single call.
```
Comment (KarhouTam) [inline docs/source/accelerator/testing.md:58]: This one is important. We should highlight this. Or developers can overlap it with no intention. I wonder will it better that we provide another api to _append_ test configs? With that, test configs can be set multiple times and is more flexible. Wdyt ? @fffrog @albanD

**Review by KarhouTam — COMMENTED** [review]
Hi @orangeH25. Thanks for your great work. I left some comments, please take a look!
